### PR TITLE
PostalCode: Add postalcode validator and alias

### LINKED ISSUE #407: Expose supplemental CLDR data
It'd be nice to have things such as [currency symobls](http://www.unicode.org/repos/cldr-aux/json/25/main/es/currencies.json), [phone codes](http://www.unicode.org/repos/cldr-aux/json/25/supplemental/telephoneCodeData.json), and probably some other supplemental CLDR data exposed by globalize.js. 

It can be as simple as a thin wrapper class that provides an API for retrieving, e.g., currency symbols and phone codes based on a country code. Just to give you an idea, [here's](https://github.com/twitter/twitter-cldr-js/tree/master/lib/twitter_cldr/js/mustache/shared) how we expose this data in twitter-cldr-js (and the specs are over [here](https://github.com/twitter/twitter-cldr-js/tree/master/spec/js/shared)).

### PR DESCRIPTION
**Pull request by mwmw7**
Partially fixes #407

### CODE REVIEW

**Comment by mwmw7** [comment]
I've added new api, unit test (functional test) please check this PR and let me know something wrong

File: src/build/intro-postalcode.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+		// Global
+		factory( root.Cldr, root.Globalize );
+	}
+}(this, function( Cldr, Globalize ) {
+
+var validateCldr = Globalize._validateCldr,
+	validateDefaultLocale = Globalize._validateDefaultLocale,
+	validateParameterPresence = Globalize._validateParameterPresence,
+    validateParameterTypeString = Globalize._validateParameterTypeString;
```
Comment (rxaviers) [inline src/build/intro-postalcode.js]: indentation with tabs

File: src/postalcode.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+ *
+ * TRUE if the value is matched with postalcode pattern given the locale.
+ * FALSE if the value is not matched with postalcode pattern given the locale.
+ *
+ * Return the boolean.
+ */
+Globalize.validatePostalCode =
+Globalize.prototype.validatePostalCode = function( value ) {
+	validateParameterPresence( value, "value" );
+    validateParameterTypeString( value, "value" );
```
Comment (rxaviers) [inline src/postalcode.js]: indentation with tabs

File: test/functional/postalcode/postalcode.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        assert.equal( Globalize( "en" ).validatePostalCode("12345-6789"), true );
+        assert.equal( Globalize( "en" ).validatePostalCode("12345-67899"), false );
+        assert.equal( Globalize( "en" ).validatePostalCode("12345 6789"), true );
+        assert.equal( Globalize( "en" ).validatePostalCode("12345 67899"), false );
+
+        assert.equal( Globalize( "ar" ).validatePostalCode("12345"), true );
+        assert.equal( Globalize( "ar" ).validatePostalCode("123456"), false );
+        assert.equal( Globalize( "ar" ).validatePostalCode("12345-12"),false );
+
+        assert.equal( Globalize( "ja" ).validatePostalCode("123-4567"), true );
```
Comment (rxaviers) [inline test/functional/postalcode/postalcode.js]: spaces
```diff
- validatePostalCode("123-4567"),
+ validatePostalCode( "123-4567" ),

File: src/postalcode.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+ * FALSE if the value is not matched with postalcode pattern given the locale.
+ *
+ * Return the boolean.
+ *
+ */
+Globalize.postalCodeValidator =
+Globalize.prototype.postalCodeValidator = function() {
+	var cldr, regex,
+        beginning = "^",
+        end = "$";
```
Comment (rxaviers) [inline src/postalcode.js]: Both are used once below. Can we get the strings inlined below and avoid the variable?

File: src/postalcode.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+	cldr = this.cldr;
+
+	validateDefaultLocale( cldr );
+
+	cldr.on( "get", validateCldr );
+    regex = new RegExp(beginning + cldr.supplemental("postalCodeData/{region}") + end);
+	cldr.off( "get", validateCldr );
```
Comment (rxaviers) [inline src/postalcode.js]: Given you call `cldr.supplemental()` once, you can simplify your code with:

```js
cldr.once( "get", validateCldr );
regex = new RegExp(beginning + cldr.supplemental("postalCodeData/{region}") + end);
```

**Comment by rxaviers** [comment]
Thanks for your contribution. Please, review your coding style and my comments above.

**Comment by rxaviers** [comment]
About the API, it looks consistent with what has been discussed in issue #407. Although, I would like to hear more opinions about it. I feel it would be correct to get the region from an argument, e.g., `postalCodeValidator( region )`, which default value is locale's region.

File: src/postalcode.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+ *
+ * Return a validatePostalCode function (of the form below).
+ *
+ * fn( value )
+ *
+ * @value [String]
+ *
+ *
+ * TRUE if the value is matched with postalcode pattern given the locale.
+ * FALSE if the value is not matched with postalcode pattern given the locale.
```
Comment (jzaefferer) [inline src/postalcode.js]: This makes it look like `FALSE` is some kind of special constant. Let's stick with `true` and `false`. Could also simplify this into a single sentence: "Returns true if the value matches the postalcode pattern for the given locale, otherwise returns false."

File: src/postalcode.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+	cldr.on( "get", validateCldr );
+    regex = new RegExp(beginning + cldr.supplemental("postalCodeData/{region}") + end);
+	cldr.off( "get", validateCldr );
+
+	return function( value ) {
+		validateParameterPresence( value, "value" );
+		validateParameterTypeString( value, "value" );
+
+		return regex.test(value);
```
Comment (jzaefferer) [inline src/postalcode.js]: Missing spaces inside parens

File: src/postalcode.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	var cldr, regex,
+        beginning = "^",
+        end = "$";
+
+	cldr = this.cldr;
+
+	validateDefaultLocale( cldr );
+
+	cldr.on( "get", validateCldr );
+    regex = new RegExp(beginning + cldr.supplemental("postalCodeData/{region}") + end);
```
Comment (jzaefferer) [inline src/postalcode.js]: Indent using tabs.

Should this set any flags, like ignore-case? Or allow the user to specify those flags?

File: test/functional/postalcode/postalcode.js [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        });
+    });
+
+    QUnit.test( "should validate CLDR content", function( assert ) {
+        util.assertCldrContent( assert, function() {
+            Globalize.validatePostalCode("427-090");
+        });
+    });
+    QUnit.test( "should return boolean", function( assert ) {
+        extraSetup();
```
Comment (jzaefferer) [inline test/functional/postalcode/postalcode.js]: This is used just once, inline the method (put the `Globalize.load` call in here directly).

File: test/functional/postalcode/postalcode.js [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        assert.equal(Globalize.validatePostalCode("427-090") ,true);
+        assert.equal(Globalize.validatePostalCode("4270-090") ,false);
+        assert.equal(Globalize.validatePostalCode("427-0090") ,false);
+        assert.equal(Globalize.validatePostalCode("427090") , false);
+        assert.equal(Globalize.validatePostalCode("thisistest") ,false);
```
Comment (jzaefferer) [inline test/functional/postalcode/postalcode.js]: Plenty bad spacing here: Missing spaces inside parens (inner and outer), invalid space in front of comma, missing space after comma.

File: test/functional/postalcode/postalcode.js [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    });
+
+    QUnit.test( "should validate CLDR content", function( assert ) {
+        util.assertCldrContent( assert, function() {
+            Globalize.validatePostalCode("427-090");
+        });
+    });
+    QUnit.test( "should return boolean", function( assert ) {
+        extraSetup();
+        assert.equal(Globalize.validatePostalCode("427-090") ,true);
```
Comment (jzaefferer) [inline test/functional/postalcode/postalcode.js]: Let's use `strictEqual` for all of these.

File: test/functional/postalcode/postalcode.js [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+            };
+        });
+    });
+
+    QUnit.test( "should validate CLDR content", function( assert ) {
+        util.assertCldrContent( assert, function() {
+            Globalize.validatePostalCode("427-090");
+        });
+    });
+    QUnit.test( "should return boolean", function( assert ) {
```
Comment (jzaefferer) [inline test/functional/postalcode/postalcode.js]: Is there a list of global sample postal codes that we could use to check more than these few handpicked examples?

Comment (mwmw7) [inline test/functional/postalcode/postalcode.js]: do you have any idea about "list of global sample postal code"?

File: src/postalcode.js (hunk 8 above)
Comment (mwmw7) [inline src/postalcode.js]: i think we do not have to care about ignore-case.

because patterns that comes from CLDR is standard format, so if someone use lower case for 
Andorra postalcode like this -> "ad123"
than it should be false 

 "AD": "AD\\d{3}", -> this is pattern from postalCodeData.json for the Andorra 
 "AR": "([A-HJ-NP-Z])?\\d{4}([A-Z]{3})?" -> this is pattern form postalCodeData for the Argentina

 
and i could not find lower case postalcode in below side
http://www.geopostcodes.com/Andorra_La_Vella -

Comment (jzaefferer) [inline src/postalcode.js]: Based on that information I think this is fine without flags and without options for flags. If someone reports an issue at some point later, we can still address it. Adding an argument to an argument-free method shouldn't be an issue.

File: src/postalcode.js [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * @value [String]
+ *
+ * Returns true if the value matches the postalcode pattern for the given locale,
+ * otherwise returns false
+ *
+ * Return the boolean.
+ *
+ */
+	Globalize.prototype.postalCodeValidator = function() {
```
Comment (rxaviers) [inline src/postalcode.js]: Misses `Globalize.postalCodeValidator =`. Also, something is weird with this indentation.

File: src/postalcode.js [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+	cldr = this.cldr;
+
+	validateDefaultLocale( cldr );
+
+	cldr.once( "get", validateCldr );
+	pattern = "^" + cldr.supplemental("postalCodeData/{region}") + "$";
+	regex = new RegExp(pattern);
+
+	return function( value, flag ) {
```
Comment (rxaviers) [inline src/postalcode.js]: Please, could you revert the `flag` option. Sorry about the back and forth.

Comment (rxaviers) [inline src/postalcode.js]: For clarity, let's go with no `flag` option.

File: test/functional.js [hunk 15]
```diff
@@ -40,7 +40,10 @@ require([
 	"./functional/number/parse-number",
 
 	// plural
-	"./functional/plural/plural"
+	"./functional/plural/plural",
+
+    // postalcode
+    "./functional/postalcode/postalcode"
```
Comment (rxaviers) [inline test/functional.js]: Still using spaces...

**Comment by rxaviers** [comment]
Please, can you change the API to accept an optional argument `region`. I believe it makes sense to allow the validation of other regions without the need of creating another instance using a locale of such region. Therefore:

- `postalCodeValidator( [region] )( value )` and 
- `validatePostalCode( value [, region] )`

The argument `region` defaults to the locale's region (`var region = options.region || this.cldr.attributes.region`).

It allows this:

```js
var enUs = new Globalize( "en-US" );
```
| zip | `enUs.validatePostalCode( zip )` | enUs.validatePostalCode( zip, "CN" ) |
| --- | --- | --- |
| "12345" | :heavy_check_mark: `true` | :heavy_multiplication_x: `false` |
| "123456" | :heavy_multiplication_x: `false` | :heavy_check_mark: `true` |
| "12345-6789" | :heavy_check_mark: `true` | :heavy_multiplication_x: `false` |

**Comment by rxaviers** [comment]
@jzaefferer please what do you think of my above comment?

File: src/postalcode.js [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+ */
+	Globalize.prototype.postalCodeValidator = function() {
+		var cldr, regex, pattern;
+
+	cldr = this.cldr;
+
+	validateDefaultLocale( cldr );
+
+	cldr.once( "get", validateCldr );
+	pattern = "^" + cldr.supplemental("postalCodeData/{region}") + "$";
```
Comment (rxaviers) [inline src/postalcode.js]: By allowing an optional region, I would consider the following changes:

```diff
- Globalize.prototype.postalCodeValidator = function() {
+ Globalize.prototype.postalCodeValidator = function( region ) {
    var cldr, regex, pattern;

    cldr = this.cldr;
+   region = region || cldr.attributes.region;

    validateDefaultLocale( cldr );

    cldr.once( "get", validateCldr );
-   pattern = "^" + cldr.supplemental("postalCodeData/{region}") + "$";
+   pattern = "^" + cldr.supplemental([ "postalCodeData", region ]) + "$";
```

File: src/postalcode.js (hunk 14 above)
Comment (mwmw7) [inline src/postalcode.js]: ok let's go with no flag option^^

File: src/postalcode.js (hunk 16 above)
Comment (mwmw7) [inline src/postalcode.js]: @rxaviers 
Don't we need to check validation of "region" ? like "validateParameterTypeString(region, "region") "?

Comment (rxaviers) [inline src/postalcode.js]: Yeap we surely do, well observed.

On Monday, March 16, 2015, hosup choi <notifications@github.com> wrote:

> In src/postalcode.js
> <https://github.com/jquery/globalize/pull/418#discussion_r26545177>:
>
> > + *
> > + * Returns true if the value matches the postalcode pattern for the given locale,
> > + * otherwise returns false
> > + *
> > + * Return the boolean.
> > + *
> > + */
> > +	Globalize.prototype.postalCodeValidator = function() {
> > +		var cldr, regex, pattern;
> > +
> > +	cldr = this.cldr;
> > +
> > +	validateDefaultLocale( cldr );
> > +
> > +	cldr.once( "get", validateCldr );
> > +	pattern = "^" + cldr.supplemental("postalCodeData/{region}") + "$";
>
> @rxaviers <https://github.com/rxaviers>
> Don't we need to check validation of "region" ? like
> "validateParameterTypeString(region, "region") "?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jquery/globalize/pull/418/files#r26545177>.
>


-- 
+55 (16) 98138-1582, +1 (415) 568-5854, skype: rxaviers
http://rafael.xavier.blog.br

File: src/postalcode.js [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+ * Returns true if the value matches the postalcode pattern for the given locale,
+ * otherwise returns false
+ *
+ * Return the boolean.
+ */
+Globalize.validatePostalCode =
+Globalize.prototype.validatePostalCode = function( value, flag ) {
+	validateParameterPresence( value, "value" );
+	validateParameterTypeString( value, "value" );
+	return this.postalCodeValidator()( value, flag );
```
Comment (jzaefferer) [inline src/postalcode.js]: Also drop the `flag` here and a few lines above.

**Comment by jzaefferer** [comment]
Can we import some samples from http://www.geopostcodes.com/data ?

**Comment by mwmw7** [comment]
@jzaefferer 
We could download bunch of dummy data in [here!](http://download.geonames.org/export/dump/)
and i am working on with these data to test our validation API
but these posalcodes what i downloaded are already verified so there is no error in it

So my question is it is meaningful test ?

**Comment by jzaefferer** [comment]
Considering that all we do is call RegExp#test, there's no real point in further testing. We could do that to check for bugs in CLDR, but until we have a reason to suspect bugs in CLDR, that's pointless.

In short: Nevermind. Let's just fix the remaining issues and land this.

**Comment by mwmw7** [comment]
@rxaviers @rxaviers i think you are so busy right now^^

File: src/postalcode.js [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+ * Return the boolean.
+ */
+Globalize.validatePostalCode =
+Globalize.prototype.validatePostalCode = function( value, region ) {
+	validateParameterPresence( value, "value" );
+	validateParameterTypeString( value, "value" );
+	return this.postalCodeValidator( region )( value );
+};
+
+Globalize.postalCodeValidator =
```
Comment (rxaviers) [inline src/postalcode.js]: Extra code

File: src/postalcode.js [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+ *
+ * Returns true if the value matches the postalcode pattern for the given locale,
+ * otherwise returns false
+ *
+ * Return the boolean.
+ *
+ */
+Globalize.postalCodeValidator =
+Globalize.prototype.postalCodeValidator = function( region ) {
+	var cldr, regex, pattern;
```
Comment (rxaviers) [inline src/postalcode.js]: Can you sort it alphabetically `cldr, pattern, regex` please?

File: test/functional/postalcode/postalcode.js [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+		assert.strictEqual( Globalize.postalCodeValidator( "US" )( "123456" ), false );
+		assert.strictEqual( Globalize.postalCodeValidator( "CN" )( "123456" ), true );
+		assert.strictEqual( Globalize.postalCodeValidator( "US" )( "12345-6789" ), true );
+		assert.strictEqual( Globalize.postalCodeValidator( "CN" )( "12345-6789" ), false );
+
+		kr.features.forEach(function( item ){
+			assert.strictEqual( Globalize( "ko" ).validatePostalCode( item.properties.Postcode ), true );
+		});
+		us.features.forEach(function( item ){
+			assert.strictEqual( Globalize( "us" ).validatePostalCode( item.properties.Postcode ), true );
```
Comment (rxaviers) [inline test/functional/postalcode/postalcode.js]: Locale should be `en-US`.

File: test/functional/postalcode/dummy/us.json [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+{ "type": "Feature", "properties": { "ISO": "US", "Country": "United States", "Language": "EN", "ID": 2004429784, "Region1": "Alabama", "Region2": "Autauga", "Region3": "", "Region4": "", "Postcode": "36003", "Locality": "Autaugaville", "Suburb": "", "Street": "S Pickett Street", "Range": "81 - 320", "Building": "", "Elevation": 49, "ISO2": "US-AL", "FIPS": "01001", "NUTS": "", "HASC": "US.AL.AU", "STAT": "", "Timezone": "America/Chicago", "UTC": "-06:00", "DST": "-05:00" }, "geometry": { "type": "Point", "coordinates": [ -86.650612, 32.430410 ] } },
+{ "type": "Feature", "properties": { "ISO": "US", "Country": "United States", "Language": "EN", "ID": 2004429785, "Region1": "Alabama", "Region2": "Autauga", "Region3": "", "Region4": "", "Postcode": "36003", "Locality": "Autaugaville", "Suburb": "", "Street": "State Route 14", "Range": "2179 - 3699", "Building": "", "Elevation": 49, "ISO2": "US-AL", "FIPS": "01001", "NUTS": "", "HASC": "US.AL.AU", "STAT": "", "Timezone": "America/Chicago", "UTC": "-06:00", "DST": "-05:00" }, "geometry": { "type": "Point", "coordinates": [ -86.701424, 32.444020 ] } }
+] }
```
Comment (rxaviers) [inline test/functional/postalcode/dummy/us.json]: Can we move these files into ./test/fixtures/postalcode?

@jzaefferer, given we are testing postalcodes intensively this way, why are we testing US and KR postalcodes only? Can we test them all?

How are we supposed to update this files?

File: test/functional/postalcode/postalcode.js [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+], function( Globalize, likelySubtags, postalcode, util, kr, us ) {
+
+	QUnit.module( ".validatePostalCode( value )", {
+		setup: function() {
+			Globalize.load( likelySubtags, {
+				main: {
+					ko: {}
+				}
+			});
+			Globalize.locale( "ko" );
```
Comment (rxaviers) [inline test/functional/postalcode/postalcode.js]: Well observed hack (using empty main ko data to allow ko bundles given we aren't actually using any ko data for postalcode manipulation). Although, the more I see how we are using postalcode methods, the more I'm convinced we should require the `region` argument and avoid using cldr's region.

File: test/functional/postalcode/dummy/us.json (hunk 21 above)
Comment (mwmw7) [inline test/functional/postalcode/dummy/us.json]: @rxaviers 
To test other countries we have to download all samples(179) json file in (here!)[http://www.geopostcodes.com/data]

File: test/functional/postalcode/postalcode.js (hunk 22 above)
Comment (mwmw7) [inline test/functional/postalcode/postalcode.js]: @rxaviers  yes right we surely need a  **region** argument

File: src/postalcode.js (hunk 18 above)
Comment (mwmw7) [inline src/postalcode.js]: @rxaviers  
Extra code? do i have remove something ?

File: doc/api/date/date-formatter.md [hunk 23]
```diff
@@ -131,6 +131,42 @@ are: `full`, `long`, `medium`, or `short`.
 
 Use skeletons for more flexibility (see its description [above](#parameters)).
 
+| `skeleton` | `Globalize( "en" ).dateFormatter( skeleton )( new Date( 2010, 10, 1, 17, 55 ) )` |
```
Comment (jzaefferer) [inline doc/api/date/date-formatter.md]: Why is this here? May need to rebase this PR.

**Comment by jzaefferer** [comment]
Let me rephrase an earlier comment, since I failed to express it properly (sorry for the back and forth): Let's not import any external text data! Its unnecessary and just adds overhead.

If we ever run into an issue with the CLDR data itself, we'll report that as a bug against CLDR. If we for some reason really need more testing on our end, we can revisit importing more fixtures.

**Comment by rxaviers** [comment]
Sounds as a good plan to me. Thanks Jörn.

File: test/functional/postalcode/postalcode.js (hunk 22 above)
Comment (rxaviers) [inline test/functional/postalcode/postalcode.js]: @mwmw7 @jzaefferer I need your input:

Considering we require the `region` argument as stated above, how about changing the API to allow the static method only `Globalize.validatePostalCode()` and `Globalize.postalCodeValidator`, but not `Globalize.prototype.validatePostalCode` and `Globalize.prototype.postalCodeValidator`? Any instance method requires initialization using a locale. But, the postalcode validation actually requires no locale data. Having postal code validation methods statically on Globalize avoids confusion.

File: test/functional/postalcode/dummy/us.json (hunk 21 above)
Comment (rxaviers) [inline test/functional/postalcode/dummy/us.json]: Given @jzaefferer comment, let's drop these fixtures please.

Comment (mwmw7) [inline test/functional/postalcode/dummy/us.json]: @rxaviers yes let's drop these

File: test/functional/postalcode/postalcode.js (hunk 22 above)
Comment (mwmw7) [inline test/functional/postalcode/postalcode.js]: @rxaviers it is good idea, but i think giving two way to use is not too bad.

and other modules are already use this alias pattern (number.js, date.js...)

File: doc/api/date/date-formatter.md (hunk 23 above)
Comment (mwmw7) [inline doc/api/date/date-formatter.md]: ok i will check it

File: test/functional/postalcode/postalcode.js (hunk 22 above)
Comment (rxaviers) [inline test/functional/postalcode/postalcode.js]: For clarity, I'm OK with keeping the alias method `Globalize.validatePostalCode()`. I'm against having the instance methods.

File: doc/api/date/date-formatter.md (hunk 23 above)
Comment (rxaviers) [inline doc/api/date/date-formatter.md]: A rebase should solve it.

File: test/functional/postalcode/postalcode.js (hunk 22 above)
Comment (mwmw7) [inline test/functional/postalcode/postalcode.js]:
```
Globalize.load
Globalize.loadMessages
...
```
These are static methods right?  
and  the method such as "load", "loadMessage" behind the "Globalize." looks like initialize  API of "Globalize". and it is make sense.
but  ```Globalize.validatePostalCode()``` is a little bit strange. 

i think most of users are trying to use like below (this snip code from [our sample!](https://github.com/mwmw7/globalize/blob/master/examples/amd-bower/main.js)
```
en = Globalize( "en" );

	// Use Globalize to format dates.
	document.getElementById( "date" ).innerHTML = en.formatDate( new Date(), {
		datetime: "medium"
	})
```
so i think we need "instance method"

Comment (rxaviers) [inline test/functional/postalcode/postalcode.js]: I'd like to hear @jzaefferer input on this...

Comment (jzaefferer) [inline test/functional/postalcode/postalcode.js]: I'm going with Rafael's suggestion, no instance methods. Less is more! We can add more later.

Comment (mwmw7) [inline test/functional/postalcode/postalcode.js]: @jzaefferer @rxaviers ok less is more ^^

**Comment by mwmw7** [comment]
i remove "dummy" path

**Comment by rxaviers** [comment]
Please, just let us know when you finish your updates.

**Comment by mwmw7** [comment]
@rxaviers  i finished updates

File: doc/api/date/date-formatter.md [hunk 24]
```diff
@@ -1,157 +0,0 @@
...
-it on each loop.
-
-```javascript
-var dates = [ new Date( a ), new Date( b ), ... ];
-var formatter = Globalize( "en" ).dateFormatter({ time: "short" });
-
-formattedDates = dates.map(function( date ) {
-  return formatter( date );
-});
-```
```
Comment (rxaviers) [inline doc/api/date/date-formatter.md]: I'm wondering why this file has been deleted?

File: src/postalcode.js [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * @value [String]
+ *
+ * Returns true if the value matches the postalcode pattern for the given locale,
+ * otherwise returns false
+ *
+ * Return the boolean.
+ */
+Globalize.validatePostalCode =
+Globalize.prototype.validatePostalCode = function( value, region ) {
```
Comment (rxaviers) [inline src/postalcode.js]: Are you going to delete the instance methods?

File: src/postalcode.js [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+ * @value [String]
+ *
+ * Returns true if the value matches the postalcode pattern for the given locale,
+ * otherwise returns false
+ *
+ * Return the boolean.
+ *
+ */
+Globalize.postalCodeValidator =
+Globalize.prototype.postalCodeValidator = function( region ) {
```
Comment (rxaviers) [inline src/postalcode.js]: Are you going to delete the instance methods?

File: test/functional/postalcode/postalcode.js [hunk 27]
```diff
@@ -0,0 +1,90 @@
+define([
+	"globalize",
+	"json!cldr-data/supplemental/likelySubtags.json",
+	"json!cldr-data/supplemental/postalCodeData.json",
+	"../../util",
+
+	"globalize/postalcode"
+
```
Comment (rxaviers) [inline test/functional/postalcode/postalcode.js]: Delete this empty line please.

File: test/functional/postalcode/postalcode.js [hunk 28]
```diff
@@ -0,0 +1,90 @@
...
+	});
+
+	QUnit.test( "should validate CLDR content", function( assert ) {
+		util.assertCldrContent( assert, function() {
+			Globalize.validatePostalCode( "427-090" );
+		});
+	});
+	QUnit.test( "should return boolean", function( assert ) {
+		Globalize.load( postalcode );
+		assert.strictEqual( Globalize.validatePostalCode( "427-090" ), true );
```
Comment (rxaviers) [inline test/functional/postalcode/postalcode.js]: How does this test pass? It misses the required `region`. Right?

File: src/postalcode.js [hunk 29]
```diff
@@ -0,0 +1,72 @@
...
+	var cldr, pattern, regex;
+
+	cldr = this.cldr;
+	region = region || cldr.attributes.region;
```
Comment (rxaviers) [inline src/postalcode.js]: We have agreed on making `region` required. This is still here. I'm confused.

File: src/postalcode.js (hunk 25 above)
Comment (mwmw7) [inline src/postalcode.js]: yes i will do that

File: doc/api/date/date-formatter.md (hunk 24 above)
Comment (mwmw7) [inline doc/api/date/date-formatter.md]: Actually this file is from @ManrajGrover, it is not mine
The problem  is my rebase command  to sync with upstream
 
i think this is not big issue 
because this file could be merged when my PR is merged to upstream/master

**Comment by rxaviers** [comment]
Bad news. CLDR has just deprecated `postalCodeData` and `postCodeRegex` in its recent version 27 release ([release notes](http://cldr.unicode.org/index/downloads/cldr-27)). More information can be found on: http://unicode.org/cldr/trac/ticket/8130.

CLDR "says" the data has not been kept up to date and there are other services that do a better job. They cite:
- ​http://i18napis.appspot.com/address/data/US
- ​http://i18napis.appspot.com/address/data/GB

The data may be removed completely in CLDR 28.

**Comment by mwmw7** [comment]
@rxaviers thanks for the news

**Comment by mwmw7** [comment]
@rxaviers  Do we have to support  only CLDR? 
i think we can finish this job if we able to use other json data.

**Comment by rxaviers** [comment]
@mwmw7 sorry for the delay. I think we would consider using another source considering it's a standard or widely adopted.

**Comment by rxaviers** [comment]
At the moment, I understand there's little chance this get implemented. So, I'm closing this issue. But, feel free to post additional comments if anyone has any input.
